### PR TITLE
Add support for Pico-ResTouch-LCD-3.5 to ili9xxx driver

### LINKED ISSUE #5324: ILI9XXX component no longer works with Waveshare Pico-ResTouch-LCD-3.5 or Waveshare 3.5inch RPi LCD (C) boards
### The problem

I had debugged and modified the ILI9XXX to work with these Waveshare colour touchscreen boards during the summer of 2023. I have these displays working flawlessly with text and graphics using the RPI PICO W, the Waveshare ESP32-S2-PICO and Waveshare ESP32-S3-PICO boards. These display boards use the ILI9488 driver with the XPT2046 touch chip. The main issue was the initialisation had width and height reversed. Also the init commands were modified to support 18bit RGB565 and reverse colours. I put these custom patches in the local my_components and every rebuild has worked flawlessly until Dec 2023. It appears major modifications for the ILI9XXX library was well as others have completely kiboshed a working configuration. I have spent days attempting to determine and debug the "release" code as well as the "dev" code, but to no avail. So now I am asking for assistance, especially from code owner @nielsnl68 on what exactly has changed and why this has caused so many issues. From the HA forum and I see numerous similar problems. 

### Which version of ESPHome has the issue?

2023.12.x

### What type of installation are you using?

Home Assistant Add-on

### Which version of Home Assistant has the issue?

2023.12.4

### What platform are you using?

ESP32

### Board

esp32-s3-devkitc-1

### Component causing the issue

ili9xxx

### Example YAML snippet

```yaml
esphome:
  name: "console"
  friendly_name: House Console
  platformio_options:
    board_build.flash_mode: dio

esp32:
  board: esp32-s3-devkitc-1
  #board: esp32s3box
  framework:
    type: esp-idf
    version: recommended
  flash_size: 16MB
  variant: ESP32S3

sensor:
  - platform: internal_temperature
    name: "CPUTemperature"
    id: esp32temp
    update_interval: 10s

color:
  - id: my_black
    red: 0%
    green: 0%
    blue: 0%


  - id: my_red
    red: 100%
    green: 0%
    blue: 0%

  - id: my_magenta
    red: 100%
    green: 0%
    blue: 100%


  - id: my_cyan
    red: 0%
    green: 100%
    blue: 100%


  - id: my_yellow
    red: 100%
    green: 100%
    blue: 0%

font:
# gfonts://family[@weight] Google Fonts
  - file: "gfonts://Roboto"
    id: roboto
    size: 20

  - file:
      type: gfonts
      family: Roboto
      weight: 900
    id: font2
    size: 16

spi:
- id: spi_1
  clk_pin: GPIO35
  mosi_pin: GPIO36
  miso_pin: GPIO37   
  #interface: hardware

display:
#Waveshare PICO 3.5 DISPLAY with ESP32-S3-PICO board.
  - platform: ili9xxx
    model: ILI9488
    spi_id: spi_1
    dc_pin: GPIO33
    cs_pin: GPIO34
    reset_pin: GPIO40
    dimensions: 480x320 
    rotation: 90
    id: ws_tft
    lambda: |-
      it.fill(my_black);
      it.print(0,120,font2, my_cyan,"my_cyan");
      it.print(0,150,font2, my_yellow,"my_yellow");
      it.print(0,180,font2, my_magenta,"my_magenta");
      it.graph(100, 20, id(temperature_graph), my_red);

graph:
  - id: temperature_graph
    duration: 30min
    width: 240
    height: 160
    traces:
      - sensor: esp32temp
        line_type: DASHED
        line_thickness: 1
        color: my_cyan
```

### Anything in the logs that might be useful for us?

_No response_

### Additional information

I am quite willing to test any bug fixes on my various platforms and displays.

### PR DESCRIPTION
**Pull request by clydebarrow**
# What does this implement/fix?

Support the Waveshare Pico-ResTouch-LCD-3.5. This is an ILI9488 with a custom SPI-parallel interface.

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes https://github.com/esphome/issues/issues/5324

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#3563

## Test Environment

- [x] ESP32
- [x] ESP32 IDF
- [ ] ESP8266
- [x] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
display:
  - platform: ili9xxx
    id: wave_disp
    model: WAVESHARE_RES_3_5
    transform:
      mirror_y: true
      swap_xy: true
    cs_pin: GPIO5
    dc_pin: GPIO22
    reset_pin: GPIO21
    spi_mode: mode0
    data_rate: 20MHz
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/ili9xxx/ili9xxx_display.h [hunk 1]
```diff
@@ -188,37 +188,37 @@ class ILI9XXXILI9488 : public ILI9XXXDisplay {
   ILI9XXXILI9488(const uint8_t *seq = INITCMD_ILI9488) : ILI9XXXDisplay(seq, 480, 320, true) {}
 
  protected:
-  void set_madctl_() override {
-    uint8_t MemoryAccessReg_Data = this->color_order_ == display::COLOR_ORDER_BGR ? MADCTL_BGR : MADCTL_RGB;
-    uint8_t DisFunReg_Data = 0x22;
+  void set_madctl() override {
+    uint8_t mad = this->color_order_ == display::COLOR_ORDER_BGR ? MADCTL_BGR : MADCTL_RGB;
```
Comment (nielsnl68) [inline esphome/components/ili9xxx/ili9xxx_display.h:192]: From the contribution page:
- function, method and variable names are lower_snake_case
- It’s preferred to use long variable/function names over short and non-descriptive ones.

The first names where better :D

**Review by nielsnl68 — CHANGES_REQUESTED** [review]
Thanks for digging into this issue and found the solution. 

I was a little shocked about the extra bit in the sending method's last night when you committed this PR,  But your explanation on the issue tread did made a lot of sense.

**Comment by nielsnl68** [comment]
@clydebarrow, You made the `data()` method 16 bit as explained [here](https://github.com/esphome/issues/issues/5324#issuecomment-1903213935). Should the `command()` method not also be 16 bits as well?

**Comment by clydebarrow** [comment]
> @clydebarrow, You made the `data()` method 16 bit as explained [here](https://github.com/esphome/issues/issues/5324#issuecomment-1903213935). Should the `command()` method not also be 16 bits as well?

No.Waveshare does not publish a schematic or describe the logic of the interface, but the sample code does exactly what I have implemented, and it works.

File: esphome/components/ili9xxx/ili9xxx_display.h (hunk 1 above)
Comment (clydebarrow) [reply esphome/components/ili9xxx/ili9xxx_display.h:192]: Well, the CI clang-tidy checks did not like the original names (I guess clang-tidy doesn't read the contribution page :-). They're only local to one small function and it's very obvious what they do, so there seems little point in tinkering with the names.

**Review by kbx81 — APPROVED** [review]

**Comment by kbx81** [comment]
@nielsnl68 I'm good with this -- are you? 😇

**Review by nielsnl68 — APPROVED** [review]
as it works then it is okay.

**Pull request merged**